### PR TITLE
refactor(error): clean-up direct error formatting (part 1)

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

See backgrounds in #11443. TL;DR: formatting errors with `{}` or `.to_string` can cause the loss of the error source. Use `thiserror_ext::Report` instead.

---

Empowered by the lints introduced in #13750.

Run `cargo install cargo-dylint dylint-link`, then add the following entry to `.vscode/settings.json` to find all places to fix:

```json
"rust-analyzer.check.overrideCommand": [
    "cargo",
    "dylint",
    "--all",
    "--workspace",
    "--",
    "--all-targets",
    "--message-format=json"
]
```

This PR fixes
```
common
batch
stream
compute
```
and some minor and related ones.

The remainings are...
```
connector
ctl
expr
frontend
jni
meta
storage (hummock, object store)
risedev
rpc client
pgwire
```

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/batch/src/executor/aggregation/orderby.rs [hunk 1]
```diff
@@ -77,7 +77,7 @@ impl ProjectionOrderBy {
     fn push_row(&self, state: &mut State, row: RowRef<'_>) -> Result<()> {
         let key =
             memcmp_encoding::encode_row(row.project(&self.order_col_indices), &self.order_types)
-                .map_err(|e| ExprError::Internal(anyhow!("failed to encode row, error: {}", e)))?;
+                .context("failed to encode row")?;
```
Comment (BugenZhao) [inline src/batch/src/executor/aggregation/orderby.rs:80]: Use `Anyhow::Context::context` to create a new internal error with the given message, while keeping the error source.

File: src/batch/src/executor/insert.rs [hunk 2]
```diff
@@ -231,8 +231,8 @@ impl BoxedExecutorBuilder for InsertExecutor {
                 .map(|IndexAndExpr { index: i, expr: e }| {
                     Ok((
                         i as usize,
-                        build_from_prost(&e.ok_or_else(|| anyhow!("expression is None"))?)
-                            .map_err(|e| anyhow!("failed to build expression: {}", e))?,
+                        build_from_prost(&e.context("expression is None")?)
```
Comment (BugenZhao) [inline src/batch/src/executor/insert.rs:234]: Besides `Result`, `context` is also applicable to `Option`s.

If you only have the error, use `anyhow::anyhow!("message").context(source)`.

File: src/batch/src/rpc/service/task_service.rs [hunk 3]
```diff
@@ -158,12 +160,12 @@ impl TaskService for BatchServiceImpl {
...
+        let mut output = task.get_task_output(&pb_task_output_id).inspect_err(|e| {
             error!(
-                "failed to get task output of Task {:?} in local execution mode",
-                task_id
+                error = %e.as_report(),
+                ?task_id,
+                "failed to get task output in local execution mode",
             );
-            e
         })?;
```
Comment (BugenZhao) [inline src/batch/src/rpc/service/task_service.rs:169]: Prefer `inspect_err` if not going to change the error variant. For example, logging it out.

File: src/batch/src/task/task_execution.rs [hunk 4]
```diff
@@ -671,7 +672,7 @@ impl<C: BatchTaskContext> BatchTaskExecution<C> {
 
         let error = error.map(Arc::new);
         *self.failure.lock() = error.clone();
-        let err_str = error.as_ref().map(|e| e.to_string());
+        let err_str = error.as_ref().map(|e| e.to_report_string());
```
Comment (BugenZhao) [inline src/batch/src/task/task_execution.rs:675]: If we really want to eagerly format the error, use `to_report_string` to replace `to_string`. This is basically an alias of `e.as_report().to_string()`.

Note that `to_string` implies `Display` trait. The behavior of `{}` `{:#}` `{:?}` `{:#?}` differ, and there're also corresponding `to_report_string` variants. Check the documentation of `Report`.

File: src/compute/src/rpc/service/config_service.rs [hunk 5]
```diff
@@ -32,9 +33,9 @@ impl ConfigService for ConfigServiceImpl {
         _request: Request<ShowConfigRequest>,
     ) -> Result<Response<ShowConfigResponse>, Status> {
         let batch_config = serde_json::to_string(self.batch_mgr.config())
-            .map_err(|e| tonic::Status::internal(format!("{}", e)))?;
+            .map_err(|e| e.to_status(Code::Internal, "compute"))?;
         let stream_config = serde_json::to_string(&self.stream_mgr.config().await)
-            .map_err(|e| tonic::Status::internal(format!("{}", e)))?;
+            .map_err(|e| e.to_status(Code::Internal, "compute"))?;
```
Comment (BugenZhao) [inline src/compute/src/rpc/service/config_service.rs:38]: Use `to_status` to convert an error to `tonic::Status`. If this is not ad-hoc, you may want to directly `impl From<..> for tonic::Status`.

File: src/error/src/tonic.rs [hunk 6]
```diff
@@ -145,6 +146,7 @@ impl std::fmt::Display for TonicStatusWrapper {
             write!(f, " to {} service", service_name)?;
         }
         write!(f, " failed: {}: ", self.0.code())?;
+        #[allow(rw::format_error)] // intentionally format the source itself
```
Comment (BugenZhao) [inline src/error/src/tonic.rs:149]: Ignore the customized lints just like how we do for `clippy`.

File: src/compute/src/server.rs [hunk 7]
```diff
@@ -428,12 +429,12 @@ pub async fn compute_node_serve(
                         _ = tokio::signal::ctrl_c() => {},
                         _ = &mut shutdown_recv => {
                             for (join_handle, shutdown_sender) in sub_tasks {
-                                if let Err(err) = shutdown_sender.send(()) {
-                                    tracing::warn!("Failed to send shutdown: {:?}", err);
+                                if let Err(_err) = shutdown_sender.send(()) {
+                                    tracing::warn!("Failed to send shutdown");
```
Comment (BugenZhao) [inline src/compute/src/server.rs:433]: No need to print the `SendError` if the value is meaningless, as it does not contain useful information.

File: src/compute/src/rpc/service/config_service.rs [hunk 8]
```diff
@@ -32,9 +33,9 @@ impl ConfigService for ConfigServiceImpl {
         _request: Request<ShowConfigRequest>,
     ) -> Result<Response<ShowConfigResponse>, Status> {
         let batch_config = serde_json::to_string(self.batch_mgr.config())
-            .map_err(|e| tonic::Status::internal(format!("{}", e)))?;
+            .map_err(|e| e.to_status(Code::Internal, "compute"))?;
```
Comment (xxchan) [inline src/compute/src/rpc/service/config_service.rs:36]: Regarding the "compute", how will the error look like finally?

File: src/compute/src/server.rs (hunk 7 above)
Comment (xxchan) [reply src/compute/src/server.rs:433]: Note: the value is the data not sent

File: src/stream/src/error.rs [hunk 9]
```diff
@@ -20,13 +20,20 @@ use risingwave_pb::PbFieldNotFound;
...
 
 /// The error type for streaming tasks.
-#[derive(thiserror::Error, Debug, thiserror_ext::Box)]
+#[derive(
+    thiserror::Error,
+    Debug,
+    thiserror_ext::Arc,
+    thiserror_ext::ContextInto,
+    thiserror_ext::Construct,
+)]
```
Comment (xxchan) [inline src/stream/src/error.rs:36]: Why `Arc`

What is `ContextInto`

**Review by xxchan — APPROVED** [review]
Others LGTM. 🫡 for the work. But I'd be even happier if the PR can be even smaller.

File: src/stream/src/error.rs (hunk 9 above)
Comment (BugenZhao) [reply src/stream/src/error.rs:36]: > Why `Arc`

The error is cloned here: https://github.com/risingwavelabs/risingwave/blob/f5efe3f3a7a7d1271cf60a1eb4f56e64c18ad569/src/stream/src/task/barrier_manager/managed_state.rs#L179-L181

> What is `ContextInto`

It's another kind of constructor derivation! 😄 Basically, we can write
https://github.com/risingwavelabs/risingwave/blob/f5efe3f3a7a7d1271cf60a1eb4f56e64c18ad569/src/stream/src/task/barrier_manager/managed_state.rs#L167-L168

as a sugar of 
```rust
let err = StreamError::unexpected_exit(err, actor_id);
```

It's not that sweet here, but can be useful for transforming the `Result`s as it's also implemented on `Result`s.

https://github.com/risingwavelabs/risingwave/blob/f5efe3f3a7a7d1271cf60a1eb4f56e64c18ad569/src/common/heap_profiling/src/jeprof.rs#L49-L52

It's highly inspired by `anyhow::Context` and [`snafu::Context`](https://docs.rs/snafu/latest/snafu/trait.ResultExt.html#tymethod.context). I'll write the docs for them ASAP.

Comment (BugenZhao) [reply src/stream/src/error.rs:36]: Oh, the context here is always evaluated. 😕 Perhaps I need to introduce a `with` method.

https://github.com/risingwavelabs/risingwave/blob/f5efe3f3a7a7d1271cf60a1eb4f56e64c18ad569/src/common/heap_profiling/src/jeprof.rs#L49-L52

File: .cargo/config.toml [hunk 10]
```diff
@@ -34,7 +34,13 @@ rustflags = [
 rustflags = [
     "--cfg",
     "tokio_unstable",
-    # uncomment the following two lines to enable `TaskLocalAlloc`
+
+    # Register the `rw` tool for referencing the custom lints in attributes like `#[allow(rw::xx)]`.
+    # This is essentially the same as `#![..]` in each crate root.
```
Comment (xxchan) [inline .cargo/config.toml:39]: Oh, can we use this to enable features? 🫣

Comment (BugenZhao) [reply .cargo/config.toml:39]: Yeah, really interesting! But it seems painful to also override `RUSTDOCFLAGS`. 😕

Comment (BugenZhao) [reply .cargo/config.toml:39]: Seems not to work for doc tests. I'll manually register it. 🤡

**Review by stdrc — APPROVED** [review]

**Pull request merged**